### PR TITLE
Fix issues in both vader and opal interval tree t that were causing issue 6524

### PR DESCRIPTION
**Pull request by hjelmn**

### CODE REVIEW

**Comment by hjelmn** [comment]
@sekifjikkatsu should fix the issue. IMB once again runs cleanly with XPMEM.

File: opal/class/opal_interval_tree.c [hunk 1]
```diff
@@ -355,31 +356,54 @@ int opal_interval_tree_insert (opal_interval_tree_t *tree, void *value, uint64_t
...
+    if (node->low < low) {
+        return 1;
+    }
+    if (node->high < high) {
+        return -1;
+    }
+    if (node->high > high) {
+        return 1;
+    }
+    if (node->data > data) {
```
Comment (gpaulsen) [inline opal/class/opal_interval_tree.c]: Is this `node->data > data` here intentional?  There are three reasons this stands out at first glace (without fully groking the code):
1. It seems strange to be comparing a `void * data` pointer with `>` operator.
2. This is new logic that did not exist before this PR.
3. There is no equivalent `<` test.

Perhaps you meant to use the `!=` operator?

Comment (hjelmn) [reply opal/class/opal_interval_tree.c]: It is intended as a tie breaker. The intent is to allow nodes with identical ranges as long as the data differs. This covers the vader usage.  I can add a comment to that effect.

**Comment by sekifjikkatsu** [comment]
I patched "https://patch-diff.githubusercontent.com/raw/open-mpi/ompi/pull/7283.diff"  to Open MPI 4.0.2.
I executed IMB-MPI1 at 32 processes on the single node  with the patched MPI library.
It was executed correctly 35times.

**Comment by sekifjikkatsu** [comment]
I patched "https://patch-diff.githubusercontent.com/raw/open-mpi/ompi/pull/7283.diff" to master too.

```
mpiexec (OpenRTE) 4.1.0a1
# Benchmarking PingPong
# Benchmarking PingPing
# Benchmarking Sendrecv
# Benchmarking Exchange
# Benchmarking Allreduce
# Benchmarking Reduce
# Benchmarking Reduce_scatter
# Benchmarking Allgather
# Benchmarking Allgatherv
# Benchmarking Gather
# Benchmarking Gatherv
# Benchmarking Scatter
# Benchmarking Scatterv
# Benchmarking Alltoall
# Benchmarking Alltoallv
# Benchmarking Bcast
# Benchmarking Barrier
# All processes entering MPI_Finalize
```

**Comment by sekifjikkatsu** [comment]
An another test program still failed with the patched MPI library.
```
tppack_honke_master.out: ../../../../../opal/mca/btl/vader/btl_vader_component.c:655: mca_btl_vader_poll_handle_frag: Assertion `NULL != xpmem_reg' failed.
[system2-cn00:00074] *** Process received signal ***
[system2-cn00:00074] Signal: Aborted (6)
[system2-cn00:00074] Signal code:  (-6)
[system2-cn00:00074] [ 0] linux-vdso.so.1(__kernel_rt_sigreturn+0x0)[0x40000005066c]
[system2-cn00:00074] [ 1] /lib64/libc.so.6(gsignal+0xac)[0x4000002f2dbc]
[system2-cn00:00074] [ 2] /lib64/libc.so.6(abort+0x110)[0x4000002e0928]
[system2-cn00:00074] [ 3] /lib64/libc.so.6(+0x2c460)[0x4000002ec460]
[system2-cn00:00074] [ 4] /lib64/libc.so.6(__assert_perror_fail+0x0)[0x4000002ec4c8]
[system2-cn00:00074] [ 5] /fefs01/mpi/seki/intratest/honke_master/master_lib_dbg/lib/openmpi/mca_btl_vader.so(mca_btl_vader_poll_handle_frag+0x110)[0x400002448d74]
[system2-cn00:00074] [ 6] /fefs01/mpi/seki/intratest/honke_master/master_lib_dbg/lib/openmpi/mca_btl_vader.so(+0x8ee0)[0x400002448ee0]
[system2-cn00:00074] [ 7] /fefs01/mpi/seki/intratest/honke_master/master_lib_dbg/lib/openmpi/mca_btl_vader.so(+0x9218)[0x400002449218]
[system2-cn00:00074] [ 8] /fefs01/mpi/seki/intratest/honke_master/master_lib_dbg/lib/libopen-pal.so.0(opal_progress+0x34)[0x4000005a829c]
[system2-cn00:00074] [ 9] /fefs01/mpi/seki/intratest/honke_master/master_lib_dbg/lib/openmpi/mca_pml_ob1.so(+0xca60)[0x4000036bca60]
[system2-cn00:00074] [10] /fefs01/mpi/seki/intratest/honke_master/master_lib_dbg/lib/openmpi/mca_pml_ob1.so(mca_pml_ob1_recv+0x36c)[0x4000036bdfcc]
[system2-cn00:00074] [11] /fefs01/mpi/seki/intratest/honke_master/master_lib_dbg/lib/libmpi.so.0(MPI_Recv+0x35c)[0x400000131434]
[system2-cn00:00074] [12] ./tppack_honke_master.out[0x400e58]
[system2-cn00:00074] [13] /lib64/libc.so.6(__libc_start_main+0xe4)[0x4000002e0d64]
[system2-cn00:00074] [14] ./tppack_honke_master.out[0x400bcc]
```

**Comment by sekifjikkatsu** [comment]
I executed this program at 2 processes on the single node.
When  "--mca btl_vader_single_copy_mechanism none" is specified, it executed correctly.

```
#include <stdio.h>
#include <stdlib.h>
#include "mpi.h"

#define D_BUFFER_SIZE 1024

int main(int argc,char *argv[])
{
  MPI_Comm comm;
  MPI_Status status;
  int rank,position,size,pack_size,tag = 1,data = 127;

  char buff[D_BUFFER_SIZE];
  char send_char,recv_char;
  int send_int,recv_int;
  long send_long,recv_long;
  double send_double,recv_double;

  MPI_Init(&argc,&argv);
  comm = MPI_COMM_WORLD;
  MPI_Comm_rank(comm,&rank);
  send_char = (char)data;
  send_int = data;
  send_long = (long)data;
  send_double = (double)data;

  MPI_Barrier(comm);
  if(rank == 0){
    position = 0;
    MPI_Pack(&send_char,1,MPI_CHAR,buff,D_BUFFER_SIZE,&position,comm);
    MPI_Pack(&send_int,1,MPI_INT,buff,D_BUFFER_SIZE,&position,comm);
    MPI_Pack(&send_long,1,MPI_LONG,buff,D_BUFFER_SIZE,&position,comm);
    MPI_Pack(&send_double,1,MPI_DOUBLE,buff,D_BUFFER_SIZE,&position,comm);
    MPI_Pack_size(D_BUFFER_SIZE,MPI_PACKED,comm,&size);
    MPI_Send(buff,D_BUFFER_SIZE,MPI_PACKED,1,tag,comm);

  } else if(rank == 1){
    position = 0;
    pack_size = D_BUFFER_SIZE;
    MPI_Recv(buff,D_BUFFER_SIZE,MPI_PACKED,0,tag,comm,&status);
    MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_char,1,MPI_CHAR,comm);
    MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_int,1,MPI_INT,comm);
    MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_long,1,MPI_LONG,comm);
    MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_double,1,MPI_DOUBLE,comm);
  }
  MPI_Barrier(comm);
  MPI_Finalize();
  if (rank == 0){
    printf("TP is OK \n");fflush(stdout);
  }
  return 0;
}
```

**Comment by sekifjikkatsu** [comment]
This test program failed same as above too with Open MPI 4.0.2 which is not patched of 7283.

> I executed this program at 2 processes on the single node.
> When  "--mca btl_vader_single_copy_mechanism none" is specified, it executed correctly.
> #include <stdio.h>
> #include <stdlib.h>
> #include "mpi.h"
> 
> #define D_BUFFER_SIZE 1024
> 
> int main(int argc,char *argv[])
> {
>   MPI_Comm comm;
>   MPI_Status status;
>   int rank,position,size,pack_size,tag = 1,data = 127;
> 
>   char buff[D_BUFFER_SIZE];
>   char send_char,recv_char;
>   int send_int,recv_int;
>   long send_long,recv_long;
>   double send_double,recv_double;
> 
>   MPI_Init(&argc,&argv);
>   comm = MPI_COMM_WORLD;
>   MPI_Comm_rank(comm,&rank);
>   send_char = (char)data;
>   send_int = data;
>   send_long = (long)data;
>   send_double = (double)data;
> 
>   MPI_Barrier(comm);
>   if(rank == 0){
>     position = 0;
>     MPI_Pack(&send_char,1,MPI_CHAR,buff,D_BUFFER_SIZE,&position,comm);
>     MPI_Pack(&send_int,1,MPI_INT,buff,D_BUFFER_SIZE,&position,comm);
>     MPI_Pack(&send_long,1,MPI_LONG,buff,D_BUFFER_SIZE,&position,comm);
>     MPI_Pack(&send_double,1,MPI_DOUBLE,buff,D_BUFFER_SIZE,&position,comm);
>     MPI_Pack_size(D_BUFFER_SIZE,MPI_PACKED,comm,&size);
>     MPI_Send(buff,D_BUFFER_SIZE,MPI_PACKED,1,tag,comm);
> 
>   } else if(rank == 1){
>     position = 0;
>     pack_size = D_BUFFER_SIZE;
>     MPI_Recv(buff,D_BUFFER_SIZE,MPI_PACKED,0,tag,comm,&status);
>     MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_char,1,MPI_CHAR,comm);
>     MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_int,1,MPI_INT,comm);
>     MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_long,1,MPI_LONG,comm);
>     MPI_Unpack(buff,D_BUFFER_SIZE,&position,&recv_double,1,MPI_DOUBLE,comm);
>   }
>   MPI_Barrier(comm);
>   MPI_Finalize();
>   if (rank == 0){
>     printf("TP is OK \n");fflush(stdout);
>   }
>   return 0;
> }

**Comment by hjelmn** [comment]
@sekifjikkatsu That program runs fine for me in a 4-core VM. Will try again on a Cray but otherwise I don't know what could be going wrong. Did you apply both patches?

**Comment by sekifjikkatsu** [comment]
I patched following two patches to Open MPI 4.0.2 and master.
https://github.com/open-mpi/ompi/pull/7283/commits/1145abc0b790f82ea25e24a3becad91ff502769c
https://github.com/open-mpi/ompi/pull/7283/commits/f86f805be1145ace46b570c5c518555b38e58ce

Does my operation right?

```
[imtofu3@-bash@13:13:10@seki@~/FX1000/Jikki/MPI/src/honke/200109/fjmpi]:git diff
diff --git a/opal/class/opal_interval_tree.c b/opal/class/opal_interval_tree.c
index e8ccda2024..110dccdacc 100644
--- a/opal/class/opal_interval_tree.c
+++ b/opal/class/opal_interval_tree.c
@@ -12,6 +12,7 @@
  *                         All rights reserved.
  * Copyright (c) 2015-2018 Los Alamos National Security, LLC. All rights
  *                         reserved.
+ * Copyright (c) 2020      Google, LLC. All rights reserved.
  * $COPYRIGHT$
  *
  * Additional copyrights may follow
@@ -39,7 +40,7 @@ static opal_interval_tree_node_t *opal_interval_tree_next (opal_interval_tree_t
                                                            opal_interval_tree_node_t *node);
 static opal_interval_tree_node_t * opal_interval_tree_find_node(opal_interval_tree_t *tree,
                                                                 uint64_t low, uint64_t high,
-                                                                bool exact, void *data);
+                                                                void *data);

 static opal_interval_tree_node_t *left_rotate (opal_interval_tree_t *tree, opal_interval_tree_node_t *x);
 static opal_interval_tree_node_t *right_rotate (opal_interval_tree_t *tree, opal_interval_tree_node_t *x);
@@ -355,31 +356,54 @@ int opal_interval_tree_insert (opal_interval_tree_t *tree, void *value, uint64_t
     return OPAL_SUCCESS;
 }

+static int opal_interval_tree_compare_node(opal_interval_tree_node_t *node, uint64_t low, uint64_t high, void *data) {
+    if ((data && node->low == low && node->high == high && node->data == data) ||
+        (!data && node->low <= low && node->high >= high)) {
+        return 0;
+    }
+    if (node->low > low) {
+        return -1;
+    }
+    if (node->low < low) {
+        return 1;
+    }
+    if (node->high < high) {
+        return -1;
+    }
+    if (node->high > high) {
+        return 1;
+    }
+    if (node->data > data) {
+        return -1;
+    }
+    return 1;
+}
+
 static opal_interval_tree_node_t *opal_interval_tree_find_interval(opal_interval_tree_t *tree, opal_interval_tree_node_t *node, uint64_t low,
-                                                                   uint64_t high, bool exact, void *data)
+                                                                   uint64_t high, void *data)
 {
     if (node == &tree->nill) {
         return NULL;
     }

-    if (((exact && node->low == low && node->high == high) || (!exact && node->low <= low && node->high >= high)) &&
-        (!data || node->data == data)) {
+    int check = opal_interval_tree_compare_node(node, low, high, data);
+    if (0 == check) {
         return node;
     }

-    if (low <= node->low) {
-        return opal_interval_tree_find_interval (tree, node->left, low, high, exact, data);
+    if (-1 == check) {
+        return opal_interval_tree_find_interval (tree, node->left, low, high, data);
     }

-    return opal_interval_tree_find_interval (tree, node->right, low, high, exact, data);
+    return opal_interval_tree_find_interval (tree, node->right, low, high, data);
 }

 /* Finds the node in the tree based on the key and returns a pointer
  * to the node. This is a bit a code duplication, but this has to be fast
  * so we go ahead with the duplication */
-static opal_interval_tree_node_t *opal_interval_tree_find_node(opal_interval_tree_t *tree, uint64_t low, uint64_t high, bool exact, void *data)
+static opal_interval_tree_node_t *opal_interval_tree_find_node(opal_interval_tree_t *tree, uint64_t low, uint64_t high, void *data)
 {
-    return opal_interval_tree_find_interval (tree, tree->root.left, low, high, exact, data);
+    return opal_interval_tree_find_interval (tree, tree->root.left, low, high, data);
 }

 void *opal_interval_tree_find_overlapping (opal_interval_tree_t *tree, uint64_t low, uint64_t high)
@@ -388,7 +412,7 @@ void *opal_interval_tree_find_overlapping (opal_interval_tree_t *tree, uint64_t
     opal_interval_tree_node_t *node;

     token = opal_interval_tree_reader_get_token (tree);
-    node = opal_interval_tree_find_node (tree, low, high, true, NULL);
+    node = opal_interval_tree_find_node (tree, low, high, NULL);
     opal_interval_tree_reader_return_token (tree, token);

     return node ? node->data : NULL;
@@ -536,7 +560,7 @@ int opal_interval_tree_delete (opal_interval_tree_t *tree, uint64_t low, uint64_
     opal_interval_tree_node_t *node;

     opal_interval_tree_write_lock (tree);
-    node = opal_interval_tree_find_node (tree, low, high, true, data);
+    node = opal_interval_tree_find_node (tree, low, high, data);
     if (NULL == node) {
         opal_interval_tree_write_unlock (tree);
         return OPAL_ERR_NOT_FOUND;
@@ -618,18 +642,23 @@ static void opal_interval_tree_insert_node (opal_interval_tree_t *tree, opal_int
     node->right = nill;

     /* find the leaf where we will insert the node */
+    int check = -1;
     while (n != nill) {
+        check = opal_interval_tree_compare_node(n, node->low, node->high, node->data);
+        /* node already exists */
+        assert (0 != check);
+
         if (n->max < node->high) {
             n->max = node->high;
         }

         parent = n;
-        n = ((node->low < n->low) ? n->left : n->right);
+        n = (-1 == check) ? n->left : n->right;
         assert (nill == n || n->parent == parent);
     }

     /* place it on either the left or the right */
-    if ((node->low < parent->low)) {
+    if (-1 == check) {
         parent->left = node;
     } else {
         parent->right = node;
diff --git a/opal/mca/btl/vader/btl_vader_xpmem.c b/opal/mca/btl/vader/btl_vader_xpmem.c
index 219c0bd5f7..5bfe8e1a64 100644
--- a/opal/mca/btl/vader/btl_vader_xpmem.c
+++ b/opal/mca/btl/vader/btl_vader_xpmem.c
@@ -5,6 +5,7 @@
  * Copyright (c) 2014      The University of Tennessee and The University
  *                         of Tennessee Research Foundation.  All rights
  *                         reserved.
+ * Copyright (c) 2020      Google, LLC. All rights reserved.
  * $COPYRIGHT$
  *
  * Additional copyrights may follow
@@ -44,8 +45,7 @@ static int vader_check_reg (mca_rcache_base_registration_t *reg, void *ctx)
 {
     vader_check_reg_ctx_t *vader_ctx = (vader_check_reg_ctx_t *) ctx;

-    if ((intptr_t) reg->alloc_base != vader_ctx->ep->peer_smp_rank ||
-        (reg->flags & MCA_RCACHE_FLAGS_PERSIST)) {
+    if ((intptr_t) reg->alloc_base != vader_ctx->ep->peer_smp_rank) {
         /* ignore this registration */
         return OPAL_SUCCESS;
     }
@@ -53,10 +53,26 @@ static int vader_check_reg (mca_rcache_base_registration_t *reg, void *ctx)
     vader_ctx->reg[0] = reg;

     if (vader_ctx->bound <= (uintptr_t) reg->bound && vader_ctx->base >= (uintptr_t) reg->base) {
-        opal_atomic_add (&reg->ref_count, 1);
+        if (0 == opal_atomic_fetch_add_32 (&reg->ref_count, 1)) {
+            /* registration is being deleted by a thread in vader_return_registration. the
+             * VMA tree implementation will block in mca_rcache_delete until we finish
+             * iterating over the VMA tree so it is safe to just ignore this registration
+             * and continue. */
+            vader_ctx->reg[0] = NULL;
+            return OPAL_SUCCESS;
+        }
         return 1;
     }

+    if (MCA_RCACHE_FLAGS_INVALID & opal_atomic_fetch_or_32(&reg->flags, MCA_RCACHE_FLAGS_INVALID)) {
+        /* another thread has already marked this registration as invalid. ignore and continue. */
+        vader_ctx->reg[0] = NULL;
+        return OPAL_SUCCESS;
+    }
+
+    /* let the caller know we found an overlapping registration that can be coalesced into
+     * the requested interval. the caller will remove the last reference and delete the
+     * registration. */
     return 2;
 }

@@ -67,8 +83,12 @@ void vader_return_registration (mca_rcache_base_registration_t *reg, struct mca_

     ref_count = opal_atomic_add_fetch_32 (&reg->ref_count, -1);
     if (OPAL_UNLIKELY(0 == ref_count && !(reg->flags & MCA_RCACHE_FLAGS_PERSIST))) {
-        mca_rcache_base_vma_delete (vma_module, reg);
-
+#if OPAL_DEBUG
+        int ret = mca_rcache_base_vma_delete (vma_module, reg);
+        assert (OPAL_SUCCESS == ret);
+#else
+        (void) mca_rcache_base_vma_delete (vma_module, reg);
+#endif
         opal_memchecker_base_mem_noaccess (reg->rcache_context, (uintptr_t)(reg->bound - reg->base));
         (void)xpmem_detach (reg->rcache_context);
         OBJ_RELEASE (reg);
@@ -100,16 +120,9 @@ mca_rcache_base_registration_t *vader_get_registation (struct mca_btl_base_endpo
     /* several segments may match the base pointer */
     rc = mca_rcache_base_vma_iterate (vma_module, (void *) base, bound - base, true, vader_check_reg, &check_ctx);
     if (2 == rc) {
-        /* remove this pointer from the rcache and decrement its reference count
-           (so it is detached later) */
-        mca_rcache_base_vma_delete (vma_module, reg);
-
-        /* start the new segment from the lower of the two bases */
-        base = (uintptr_t) reg->base < base ? (uintptr_t) reg->base : base;
-
-        /* remove the last reference to this registration */
-        vader_return_registration (reg, ep);
-
+        bound = bound < (uintptr_t) reg->bound ? (uintptr_t) reg->bound : bound;
+        base = base > (uintptr_t) reg->base ? (uintptr_t) reg->base : base;
+        vader_return_registration(reg, ep);
         reg = NULL;
     }

@@ -151,13 +164,16 @@ mca_rcache_base_registration_t *vader_get_registation (struct mca_btl_base_endpo
     return reg;
 }

+struct vader_cleanup_reg_ctx {
+    mca_btl_vader_endpoint_t *ep;
+    opal_list_t *registrations;
+};
+
 static int mca_btl_vader_endpoint_xpmem_rcache_cleanup (mca_rcache_base_registration_t *reg, void *ctx)
 {
-    mca_btl_vader_endpoint_t *ep = (mca_btl_vader_endpoint_t *) ctx;
-    if ((intptr_t) reg->alloc_base == ep->peer_smp_rank) {
-        /* otherwise dereg will fail on assert */
-        reg->ref_count = 0;
-        OBJ_RELEASE(reg);
+    struct vader_cleanup_reg_ctx *cleanup_ctx = (struct vader_cleanup_reg_ctx *) ctx;
+    if ((intptr_t) reg->alloc_base == cleanup_ctx->ep->peer_smp_rank) {
+        opal_list_append(cleanup_ctx->registrations, &reg->super.super);
     }

     return OPAL_SUCCESS;
@@ -165,11 +181,22 @@ static int mca_btl_vader_endpoint_xpmem_rcache_cleanup (mca_rcache_base_registra

 void mca_btl_vader_xpmem_cleanup_endpoint (struct mca_btl_base_endpoint_t *ep)
 {
+    mca_rcache_base_registration_t *reg;
+    opal_list_t registrations;
+    struct vader_cleanup_reg_ctx cleanup_ctx = {.ep = ep, .registrations = &registrations};
+
+    OBJ_CONSTRUCT(&registrations, opal_list_t);
+
     /* clean out the registration cache */
     (void) mca_rcache_base_vma_iterate (mca_btl_vader_component.vma_module,
                                         NULL, (size_t) -1, true,
                                         mca_btl_vader_endpoint_xpmem_rcache_cleanup,
-                                        (void *) ep);
+                                        (void *) &cleanup_ctx);
+    while (NULL != (reg = (mca_rcache_base_registration_t *) opal_list_remove_first(&registrations))) {
+        vader_return_registration (reg, ep);
+    }
+    OBJ_DESTRUCT(&registrations);
+
     if (ep->segment_base) {
         xpmem_release (ep->segment_data.xpmem.apid);
         ep->segment_data.xpmem.apid = 0;
```

**Comment by sekifjikkatsu** [comment]
xpmem_attach returns -1.
value of bound or base is wrong?

```
SEKI vader_get_registation PASS-1 bound=1000000000001 base=ffffffe00000
tppack_honke.out: ../../../../../opal/mca/btl/vader/btl_vader_component.c:654: mca_btl_vader_poll_handle_frag: Assertion `NULL != xpmem_reg' failed.
```

```
@@ -123,6 +136,7 @@ mca_rcache_base_registration_t *vader_get_registation (struct mca_btl_base_endpo
             reg->flags = flags;
             reg->alloc_base = (void *) (intptr_t) ep->peer_smp_rank;

+memset((void *)&xpmem_addr,0,sizeof(xpmem_addr_t));
 #if defined(HAVE_SN_XPMEM_H)
             xpmem_addr.id     = ep->segment_data.xpmem.apid;
 #else
@@ -133,6 +147,7 @@ mca_rcache_base_registration_t *vader_get_registation (struct mca_btl_base_endpo
             reg->rcache_context = xpmem_attach (xpmem_addr, bound - base, NULL);
             if (OPAL_UNLIKELY((void *)-1 == reg->rcache_context)) {
                 OBJ_RELEASE(reg);
+fprintf(stderr,"SEKI %s PASS-1 bound=%lx base=%lx \n",__func__,bound,base);fflush(stderr);
                 return NULL;
             }
```

**Comment by hjelmn** [comment]
Yup. Those do not look correct to me. Will see if I can get the reproducer failing on an XC and take a look.

**Comment by hppritcha** [comment]
bot:ompi:retest

**Comment by sekifjikkatsu** [comment]
- case that program ends normally.

```
SEKI vader_get_registation rc=0 ep=c006860 apid=27567 reg=0 base=ffffd3a00000 bound=ffffd3c00001 rem_ptr=ffffd3a8b440 size=1024 attach_align=2097152
```


- case that program fails.

```
SEKI vader_get_registation rc=0 ep=592eb0 apid=24582 reg=0 base=ffffffe00000 bound=1000000000001 rem_ptr=ffffffffda90 size=1024 attach_align=2097152
```

```
@@ -99,22 +119,18 @@ mca_rcache_base_registration_t *vader_get_registation (struct mca_btl_base_endpo

     /* several segments may match the base pointer */
     rc = mca_rcache_base_vma_iterate (vma_module, (void *) base, bound - base, true, vader_check_reg, &check_ctx);
+fprintf(stderr,"SEKI %s rc=%d ep=%lx apid=%d reg=%lx base=%lx bound=%lx rem_ptr=%lx size=%d attach_align=%ld \n",__func__,rc,ep,ep->segment_data.xpmem.apid,reg,base,bound,rem_ptr,size,attach_align);fflush(stderr);
```

**Comment by sekifjikkatsu** [comment]
rem_ptr is adress of an auto variable array (char buff[D_BUFFER_SIZE];) on the test program.

```
SEKI vader_get_registation rc=0 ep=592eb0 apid=20820 reg=0 base=400002400000 bound=400003600001 rem_ptr=400002530000 size=16777216 attach_align=2097152
SEKI vader_get_registation ep=592eb0 reg=59cc90 OBJ_NEW
TP ../honke_master/tppack.c buff=ffffffffdaa0
SEKI vader_get_registation rc=0 ep=592eb0 apid=20820 reg=0 base=ffffffe00000 bound=1000000000001 rem_ptr=ffffffffdaa0 size=1024 attach_align=2097152
SEKI vader_get_registation ep=592eb0 reg=562230 OBJ_NEW
SEKI vader_get_registation ep=592eb0 reg=562230 base=ffffffe00000 bound=1000000000001 apid=20820 FAIL-ATTACH
tppack_honke.out: ../../../../../opal/mca/btl/vader/btl_vader_component.c:654: mca_btl_vader_poll_handle_frag: Assertion `NULL != xpmem_reg' failed.
```

**Comment by hjelmn** [comment]
@sekifjikkatsu Can you give me the configure line and command used to compile your reproducer? I can't seem to trigger the error.

**Comment by sekifjikkatsu** [comment]
I builded Open MPI 4.0.2 as follows.

```
#!/bin/bash

AS=aarch64-linux-gnu-as
export AS
LD=aarch64-linux-gnu-ld
export LD
NM=aarch64-linux-gnu-nm
export NM
AR=aarch64-linux-gnu-ar
export AR
RANLIB=aarch64-linux-gnu-ranlib
export RANLIB
CC=aarch64-linux-gnu-gcc
export CC
CXX=aarch64-linux-gnu-g++
export CXX
FC=aarch64-linux-gnu-gfortran
export FC
CFLAGS=-funwind-tables
export CFLAGS
CXXFLAGS=-funwind-tables
export CXXFLAGS

echo $CC

## libtool (GNU libtool) 2.4.6
## autoconf (GNU Autoconf) 2.69
## automake (GNU automake) 1.15
## m4 (GNU M4) 1.4.17
PATH=/home/share/mpi/autotools-ompi-4.0/bin:$PATH
export PATH

## cross-compilation environment for aarch64 target
PATH=/opt/FJSVxos/devkit/aarch64/bin:${PATH}
export PATH

LD_LIBRARY_PATH=/home/share/mpi/autotools-ompi-4.0/lib:$LD_LIBRARY_PATH
export LD_LIBRARY_PATH

JDKPATH="/home/share/mpi/JDK/9"

#../configure -prefix=/home/seki/FX1000/Jikki/MPI/src/honke/200109/lib_dbg \
#   --host=aarch64-unknown-linux-gnu \
#   --enable-debug \
#   --with-xpmem=/home/share/mpi/headers/xpmem/FX1000/latest \
#   --disable-mpi-java \
#   --disable-mpi-fortran \
#   --disable-mpi-cxx \
#   --disable-mpi-io

#make clean
make
make install
```

**Comment by hjelmn** [comment]
Ok, that is the difference. I was testing on x86_64. Will try it on an AArch64 system.

**Comment by hjelmn** [comment]
Ok, I see the problem. It is unrelated to this issue. The VADER_MAX_ADDR is wrong for AARCH64. Should be 0xfffffffffffful. I will create a commit to update the value tomorrow.

**Comment by hppritcha** [comment]
bot:ompi:retest

**Comment by hppritcha** [comment]
@hjelmn did you try to push a commit that didn't go through?

**Comment by hjelmn** [comment]
@hppritcha I added a commit to get some test code to a lanl system quickly. Immediately deleted the commit. Seems to have confused CI.

**Comment by hjelmn** [comment]
I am trying to figure out the best way to get at TASK_SIZE right now. Turns out there are multiple possible values for AARCH64. Doesn't affect this PR though.

**Comment by hjelmn** [comment]
Not sure why the build checker is marked as failed. Looked through it and everything looks fine to me.

**Comment by hjelmn** [comment]
Except for the aborted by @bwbarrett ....

**Comment by hjelmn** [comment]
Looks like just a timeout. Merging.

**Comment by hppritcha** [comment]
yes doing so.

**Pull request merged**